src/solver/gemm.cpp

```diff
@@ -0,0 +1,1080 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2020 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/solver.hpp>
+
+#include <miopen/algorithm.hpp>
+#include <miopen/env.hpp>
+#include <miopen/gemm_v2.hpp>
+#include <miopen/handle.hpp>
+#include <miopen/kernel.hpp>
+#include <miopen/tensor.hpp>
+#include <miopen/tensor_ops.hpp>
+#include <miopen/util.hpp>
+#include <miopen/conv/data_invoke_params.hpp>
+
+#include <boost/any.hpp>
+#include <boost/range/adaptors.hpp>
+
+MIOPEN_DECLARE_ENV_VAR(MIOPEN_CONV_PRECISE_ROCBLAS_TIMING)
+
+namespace miopen {
+namespace solver {
+
+#if MIOPEN_USE_GEMM
+#ifdef CPPCHECK
+// Keep the value unknown in cppcheck since this can differ between opencl and hip
+static bool IsUseRocBlas;
+#else
+static constexpr const bool IsUseRocBlas = (MIOPEN_USE_ROCBLAS == 1);
+#endif
+
+static inline bool IsAnyBufferBF16(const TensorDescriptor& xDesc,
+                                   const TensorDescriptor& yDesc,
+                                   const TensorDescriptor& wDesc)
+{
+    return xDesc.GetType() == miopenBFloat16 || yDesc.GetType() == miopenBFloat16 ||
+           wDesc.GetType() == miopenBFloat16;
+}
+#endif
+
+bool GemmFwdBase::IsApplicable(const ExecutionContext&,
+                               const conv::ProblemDescription& problem) const
+{
+#if MIOPEN_USE_GEMM
+    const auto& xDesc = problem.GetIn();
+    const auto& wDesc = problem.GetWeights();
+    const auto& yDesc = problem.GetOut();
+    return problem.GetDirection() == conv::Direction::Forward &&
```

Comment:
Please also add check that layout is NCHW. This is for #744.

Reply:
https://github.com/ROCmSoftwarePlatform/MIOpen/blob/e0ffe166d7911ac6097c7d000e7a60ef126bd4f4/src/include/miopen/problem_description.hpp#L116